src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -93,6 +108,173 @@ internal ModelPath WorksharingModelPath
             get { return this.InternalDocument.GetWorksharingCentralModelPath(); }
         }
 
+        /// <summary>
+        /// Purge unused Elements from the model.
+        /// </summary>
+        /// <param name="fullPurge">Similar to clicking the purge button multiple times in Revit, the node will repeatedly purge until there is nothing left to remove.</param>
+        /// <returns>Purged element ids</returns>
+        public List<int> PurgeUnused(bool fullPurge = false)
+        {
+            List<ElementId> purgedElementIds = new List<ElementId>();
+            TransactionManager.Instance.EnsureInTransaction(this.InternalDocument);
+
+            purgedElementIds.AddRange(PurgeElements(fullPurge));
+            purgedElementIds.AddRange(PurgeMaterials());
+
+            TransactionManager.Instance.TransactionTaskDone();
+
+            if (purgedElementIds.Count == 0)
+                throw new InvalidOperationException(Properties.Resources.NoElementsToPurge);
+
+            return purgedElementIds.Select(x => x.IntegerValue).ToList();
+        }
+
+        private List<ElementId> PurgeMaterials()
+        {
+            List<Autodesk.Revit.DB.ElementId> materials = new FilteredElementCollector(this.InternalDocument)
+                .OfClass(typeof(Autodesk.Revit.DB.Material))
+                .ToElementIds()
+                .ToList();
+
+            if (materials == null)
+                return new List<ElementId>();
+
+            List<Autodesk.Revit.DB.Element> elementTypes = new FilteredElementCollector(this.InternalDocument)
+                .WhereElementIsElementType()
+                .ToElements()
+                .ToList();
+
+            if (elementTypes == null)
+                return new List<ElementId>();
+
+            var nonPurgedElements = new HashSet<ElementId>();
+
+            for (int i = 0; i < elementTypes.Count; i++)
+            {
+                nonPurgedElements.UnionWith(GetNonPurgableMaterialIds(elementTypes[i]));
```

Comment:
do we need to use `UnionWith` or can a simple `AddRange` do ? if there is a distinction between the 2, i'd add it as a comment explaining why it's needed

Reply:
`HashSet` doesn't have `AddRange`, `UnionWith` is the equivalent but works slightly different as it will only add unique elements from the new collection. 

https://stackoverflow.com/questions/15267034/is-there-an-addrange-equivalent-for-a-hashset-in-c-sharp